### PR TITLE
Improve APIs for controller startup on Darwin.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
* Adds a way of starting up a controller with an intermediate certificate.

* Allows controlling whether controller startup reuses existing certificates or
  creates new ones, and whether the operational identity or operational keys
  change in the process.  In particular, allows starting a controller with the
  same operatonal keys as it was using last time.

* Removes framework-internal writes of node id to storage, apart from whatever
  writes the fabric table does.

* Removes framework-internal writes of root certificates, apart from whatever
  writes the fabric table does.

* Removes use of the kinda-broken PERSISTENT_KEY_OP macros from Darwin code.

#### Problem
See above.

#### Change overview
See above.

#### Testing
* Lots of unit tests added.
* Ran chip-tool-darwin for some basic commands.
* Commissioned mac all-clusters-app from iOS CHIPTool and made sure I can restart CHIPTool and have things keep working.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]
Fast tracking platform changes.

File: src/darwin/Framework/CHIP/MTRCertificates.h [hunk 1]
```diff
@@ -68,12 +68,20 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
  * Check whether the given keypair's public key matches the given certificate's
- * public key.
+ * public key.  The certificate is epected to be an X.509 DER encoded
+ * certificate.
  *
  * Will return NO on failures to extract public keys from the objects.
  */
 + (BOOL)keypairMatchesCertificate:(NSData *)certificate keypair:(id<CHIPKeypair>)keypair;
```
Comment (anush-apple) [inline src/darwin/Framework/CHIP/MTRCertificates.h:76]:
```suggestion
+ (BOOL)keypair:(id<CHIPKeypair>)keypair matchesCertificate:(NSData *)certificate;
```

File: src/darwin/Framework/CHIP/MTRCertificates.h [hunk 2]
```diff
@@ -68,12 +68,20 @@ NS_ASSUME_NONNULL_BEGIN
...
 
+/**
+ * Check whether two X.509 DER encoded certificates are equivalent, in the sense
+ * of having the same public key and the same subject DN.  Returns NO if public
+ * keys or subject DNs cannot be extracted from the certificates.
+ */
++ (BOOL)isEquivalent:(NSData *)certificate1 to:(NSData *)certificate2;
```
Comment (anush-apple) [inline src/darwin/Framework/CHIP/MTRCertificates.h:83]:
```suggestion
+ (BOOL)isCertificate:(NSData *)certificate1 equalTo:(NSData *)certificate2;
```

File: src/darwin/Framework/CHIP/CHIPDeviceController.h [hunk 3]
```diff
@@ -115,9 +115,10 @@ typedef void (^CHIPDeviceConnectionCallback)(CHIPDevice * _Nullable device, NSEr
 + (instancetype)new NS_UNAVAILABLE;
 
 /**
- * Return the Node Id assigned to the controller.
+ * Return the Node Id assigned to the controller.  Will return nil if the
+ * controller is not running (and hence does not know its node id).
  */
-- (NSNumber *)getControllerNodeId;
+- (nullable NSNumber *)getControllerNodeId;
```
Comment (anush-apple) [inline src/darwin/Framework/CHIP/CHIPDeviceController.h:121]: nit pick, do we make this a readonly property?

File: src/darwin/Framework/CHIP/MTRCertificates.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRCertificates.h:76]: Yes, makes sense.

File: src/darwin/Framework/CHIP/MTRCertificates.h (hunk 2 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRCertificates.h:83]: Also makes sense.

File: src/darwin/Framework/CHIP/CHIPDeviceController.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/CHIPDeviceController.h:121]: Done.

**Pull request merged**